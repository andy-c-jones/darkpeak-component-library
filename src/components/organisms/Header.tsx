import { Button } from '../atoms/Button';
import { SiteLogo } from '../atoms/SiteLogo';
import { MainNav } from '../molecules/MainNav';

type User = {
  name: string;
};

interface HeaderProps {
  user?: User;
  logoConfig: {
    href?: string;
    src: string;
    alt: string;
  };
  navItems: {
    label: string;
    href: string;
  }[];
  onLogin: () => void;
  onLogout: () => void;
  onCreateAccount: () => void;
}

export const Header = ({
  user,
  logoConfig,
  navItems,
  onLogin,
  onLogout,
  onCreateAccount
}: HeaderProps) => (
  <header>
    <div className="flex items-center gap-12 justify-between">
      <SiteLogo href={logoConfig.href || '/'} src={logoConfig.src} alt={logoConfig.alt} />
      <MainNav navItems={navItems} />
      <div className="flex items-center gap-2 ">
        {user ? (
          <>
            <span className="welcome">
              Welcome, <b>{user.name}</b>!
            </span>
            <Button size="small" onClick={onLogout} label="Log out" />
          </>
        ) : (
          <>
            <Button size="small" onClick={onLogin} label="Log in" />
            <Button type="primary" size="small" onClick={onCreateAccount} label="Sign up" />
          </>
        )}
      </div>
    </div>
  </header>
);
