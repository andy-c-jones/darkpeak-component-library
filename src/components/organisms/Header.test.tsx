/**
 * @vitest-environment jsdom
 */
import { render, cleanup } from '@testing-library/react';
import { describe, it, expect, beforeEach, vi } from 'vitest';
import { Header } from './Header';
import userEvent from '@testing-library/user-event';

beforeEach(() => {
  cleanup();
});

describe('Header', () => {
  it('renders without error', () => {
    const { asFragment } = render(
      <Header
        logoConfig={{ src: 'test', alt: 'test' }}
        navItems={[]}
        onLogin={() => {}}
        onLogout={() => {}}
        onCreateAccount={() => {}}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the navigation items', () => {
    const c = render(
      <Header
        logoConfig={{ src: 'test', alt: 'test' }}
        navItems={[
          { label: 'Home', href: '/' },
          { label: 'About', href: '/about' }
        ]}
        onLogin={() => {}}
        onLogout={() => {}}
        onCreateAccount={() => {}}
      />
    );

    let navItems = c.container.querySelectorAll('li');

    expect(navItems.length).toBe(2);

    expect(navItems[0].querySelector('a')?.textContent).toBe('Home');
    expect(navItems[0].querySelector('a')?.getAttribute('href')).toBe('/');
    expect(navItems[1].querySelector('a')?.textContent).toBe('About');
    expect(navItems[1].querySelector('a')?.getAttribute('href')).toBe('/about');
  });

  it('renders the user name when logged in', () => {
    const c = render(
      <Header
        logoConfig={{ src: 'test', alt: 'test' }}
        navItems={[]}
        user={{ name: 'test' }}
        onLogin={() => {}}
        onLogout={() => {}}
        onCreateAccount={() => {}}
      />
    );

    expect(c.container.querySelector('.welcome')?.textContent).toBe('Welcome, test!');
  });

  it('renders the login button when not logged in', () => {
    const c = render(
      <Header
        logoConfig={{ src: 'test', alt: 'test' }}
        navItems={[]}
        onLogin={() => {}}
        onLogout={() => {}}
        onCreateAccount={() => {}}
      />
    );

    expect(c.getAllByRole('button').filter((b) => b.textContent === 'Log in').length).toBe(1);
  });

  it('renders the logout button when logged in', () => {
    const c = render(
      <Header
        logoConfig={{ src: 'test', alt: 'test' }}
        navItems={[]}
        user={{ name: 'test' }}
        onLogin={() => {}}
        onLogout={() => {}}
        onCreateAccount={() => {}}
      />
    );

    expect(c.getAllByRole('button').filter((b) => b.textContent === 'Log out').length).toBe(1);
  });

  it('renders the create account button when not logged in', () => {
    const c = render(
      <Header
        logoConfig={{ src: 'test', alt: 'test' }}
        navItems={[]}
        onLogin={() => {}}
        onLogout={() => {}}
        onCreateAccount={() => {}}
      />
    );

    console.log(c.container.innerHTML);
    expect(c.getAllByRole('button').filter((b) => b.textContent === 'Sign up').length).toBe(1);
  });

  it('does not render the create account button when logged in', () => {
    const c = render(
      <Header
        logoConfig={{ src: 'test', alt: 'test' }}
        navItems={[]}
        user={{ name: 'test' }}
        onLogin={() => {}}
        onLogout={() => {}}
        onCreateAccount={() => {}}
      />
    );

    expect(c.getAllByRole('button').filter((b) => b.textContent === 'Sign Up').length).toBe(0);
  });

  it('calls the onLogin callback when the login button is clicked', async () => {
    const onLogin = vi.fn();

    const c = render(
      <Header
        logoConfig={{ src: 'test', alt: 'test' }}
        navItems={[]}
        onLogin={onLogin}
        onLogout={() => {}}
        onCreateAccount={() => {}}
      />
    );

    await userEvent.click(c.getByText('Log in'));

    expect(onLogin).toHaveBeenCalled();
  });

  it('calls the onLogout callback when the logout button is clicked', async () => {
    const onLogout = vi.fn();

    const c = render(
      <Header
        logoConfig={{ src: 'test', alt: 'test' }}
        navItems={[]}
        user={{ name: 'test' }}
        onLogin={() => {}}
        onLogout={onLogout}
        onCreateAccount={() => {}}
      />
    );

    await userEvent.click(c.getByText('Log out'));

    expect(onLogout).toHaveBeenCalled();
  });

  it('calls the onCreateAccount callback when the create account button is clicked', async () => {
    const onCreateAccount = vi.fn();

    const c = render(
      <Header
        logoConfig={{ src: 'test', alt: 'test' }}
        navItems={[]}
        onLogin={() => {}}
        onLogout={() => {}}
        onCreateAccount={onCreateAccount}
      />
    );

    await userEvent.click(c.getByText('Sign up'));

    expect(onCreateAccount).toHaveBeenCalled();
  });
});
