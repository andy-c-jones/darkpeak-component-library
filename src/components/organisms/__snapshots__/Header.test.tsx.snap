// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > renders without error 1`] = `
<DocumentFragment>
  <header>
    <div
      class="flex items-center gap-12 justify-between"
    >
      <a
        href="/"
      >
        <img
          alt="test"
          src="test"
        />
      </a>
      <nav
        class="flex gap-6 items-center flex-wrap"
      >
        <ul
          class="flex gap-10 grow-0 flex-wrap"
        />
      </nav>
      <div
        class="flex items-center gap-2 "
      >
        <button
          class="rounded-lg w-24 h-8 text-white bg-primary-default hover:bg-primary-accent"
          type="button"
        >
          Log in
        </button>
        <button
          class="rounded-lg w-24 h-8 text-white bg-primary-default hover:bg-primary-accent"
          type="button"
        >
          Sign up
        </button>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
